<?xml version="1.0"?>
<!-- This is a launch file to bringup Kfold cross validation for hand crafted descriptors-->
<launch>
<!-- this flag will configure the system to run with nodes or nodelets-->
<arg name="use_nodelets" default="0"/>
<arg name="use_standalone" default="false"/>
<!--<arg name="use_pr2" default="0"/>-->
<arg name="pdb_source" default="default_pdb"/>

<group ns="/perception">

	<!-- _________________________________-->
	<!--|                                 |-->
	<!--|             Memory 		      |-->
	<!--|_________________________________|-->

	<!--Launches a race_perception_db node(let) -->
	<include file="$(find race_perception_db)/launch/bringup.launch">
		<arg name="use_nodelets" value="$(arg use_nodelets)" />
		<arg name="pdb_source" value="$(arg pdb_source)" />
		<arg name="name" value="perception_db" />
	</include>

	<!-- Launches the nodelet_manager -->
	<group if="$(arg use_nodelets)">
	    <node pkg="nodelet" type="nodelet" name="nodelet_manager" args="manager" output="screen" required="true">
		<param name="num_worker_threads" value="10"/>
	    </node>
	</group>
	
	<group ns="/pipeline1">
		
	<!-- _________________________________-->
	<!--|                                 |-->
	<!--|         Global Param            |-->
	<!--|_________________________________|-->

		<!-- name of approach -->
		<arg name="name_of_approach" default="COR_2021_HC" />
		<param name="/perception/name_of_approach" type="string" value = "$(arg name_of_approach)"/>

		<!-- ****************************************  -->
		<!--       set dataset_path parameter          -->
		<!-- ****************************************  -->

			<!-- restaurant_object_dataset = "$(env HOME)/datasets/restaurant_object_dataset/"; -->
			<!-- washington_RGBD_object : "$(env HOME)/datasets/washington_RGBD_object/"; -->

			<arg name="dataset_path" default="$(env HOME)/datasets/restaurant_object_dataset/" />				
			<param name="/perception/dataset_path" type="string" value = "$(arg dataset_path)"/>


		<!-- ****************************************  -->
		<!--       object descriptor parameters        -->
		<!-- ****************************************  -->
		
			<!-- object_descriptor: possible options are including [GOOD, ESF, VFH, GRSD] -->
			<arg name="object_descriptor" default="GOOD" />
			<param name="/perception/object_descriptor" type="string" value = "$(arg object_descriptor)"/>


			<!-- GOOD descriptor params -->
			<arg name="number_of_bins" default="15" />
			<param name="/perception/number_of_bins" type="int" value = "$(arg number_of_bins)"/>

			<!-- VFH descriptor params -->
			<arg name="normal_estimation_radius" default="0.03" />
			<param name="/perception/normal_estimation_radius" type="double" value = "$(arg normal_estimation_radius)"/>

		
		<!-- ****************************************  -->
		<!--       		distance functions		       -->
		<!-- ****************************************  -->
		
			<!-- possible options are : chiSquared, kLDivergance, motyka,  divergence,  euclidean, intersection, manhattan,
			cosine, dice, bhattacharyya, sorensen, canberra, pearson, neyman, gower -->
			<arg name="distance_function" default="motyka" />
			<param name="/perception/distance_function" type="string" value = "$(arg distance_function)"/>
			

		<!-- ****************************************  -->
		<!--       		KNN parameters 			       -->
		<!-- ****************************************  -->

			<!-- K param for KNN, select an odd number [1, 3, 5, 7, 9, ...]-->
			<arg name="K_for_KNN" default="1" />
			<param name="/perception/K_for_KNN" type="int" value = "$(arg K_for_KNN)"/>


		<!-- ****************************************  -->
		<!--       running_a_bunch_of_experiments 	   -->
		<!-- ****************************************  -->

			<!-- pdb_loaded 0 = FLASE, 1 = TRUE; we do not need to do training when we pass a pdb-->
			<arg name="pdb_loaded" default="0" /> 
			<param name="/perception/pdb_loaded" type="bool" value = "$(arg pdb_loaded)"/>

			<!-- running_a_bunch_of_experiments 0 = FLASE, 1 = TRUE; we do not need to do training when we pass a pdb-->
			<arg name="running_a_bunch_of_experiments" default="0" /> 
			<param name="/perception/running_a_bunch_of_experiments" type="bool" value = "$(arg running_a_bunch_of_experiments)"/>




		<!-- ________________________________-->
		<!--|                               |-->
		<!--|   DO NOT CHANGE THESE PARAMS  |-->
		<!--|_______________________________|-->
				
		<arg name="adaptive_support_lenght" default="0" />
		<param name="/perception/adaptive_support_lenght" type="int" value = "$(arg adaptive_support_lenght)"/>
		
		<arg name="sign" default="1" />
		<param name="/perception/sign" type="int" value = "$(arg sign)"/>

		<arg name="global_image_width" default="0.2" />
		<param name="/perception/global_image_width" type="double" value = "$(arg global_image_width)"/>
				
		<arg name="off_line_flag" default="0" />
		<param name="/perception/off_line_flag" type="int" value = "$(arg off_line_flag)"/>

		<arg name="recognition_threshold" default="2000" />
		<param name="/perception/recognition_threshold" type="double" value = "$(arg recognition_threshold)"/>

	<!-- ________________________________________-->
	<!--|                                       |-->
	<!--|   kfold_cross_validation_evaluation   |-->
	<!--|_______________________________________|-->
			
		<include file="$(find rug_kfold_cross_validation)/launch/bringup_hand_crafted_descriptor.launch">
			<arg name="use_nodelets" value="$(arg use_nodelets)" />
			<arg name="name" value="cross_validation" />
		</include>		
	
	</group>
	
    </group>

</launch>


